### PR TITLE
SONAR-8817 Add SonarQube version safeguard in cluster mode

### PR DESCRIPTION
**Pull request by ehartmann**

### CODE REVIEW

File: server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  public static String getSonarqubeVersion() {
+    if (sonarqubeVersion == null) {
+      loadVersions();
+    }
+    return sonarqubeVersion;
+  }
+
+  private synchronized static void loadVersions() {
```
Comment (SonarTech) [inline server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java]: MINOR Reorder the modifiers to comply with the Java Language Specification. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AModifiersOrderCheck)

File: server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ */
+
+package org.sonar.application.config;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+
+public class SonarQubeVersionHelper {
+  private static final String SONARQUBE_VERSION_PATH = "/sonarqube-version.txt";
```
Comment (SonarTech) [inline server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java]: MINOR Refactor your code to get this URI from a customizable parameter. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1075)

File: server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  public static String getSonarqubeVersion() {
+    if (sonarqubeVersion == null) {
+      loadVersions();
+    }
+    return sonarqubeVersion;
+  }
+
+  private synchronized static void loadVersions() {
+    try {
+      try (BufferedReader in = new BufferedReader(new InputStreamReader(SonarQubeVersionHelper.class.getResourceAsStream(SONARQUBE_VERSION_PATH)))) {
```
Comment (SonarTech) [inline server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java]: MAJOR Remove this use of constructor "InputStreamReader(InputStream)" [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1943)

File: server/sonar-process-monitor/src/main/java/org/sonar/application/cluster/AppStateClusterImpl.java [hunk 4]
```diff
@@ -43,9 +43,13 @@
 import org.sonar.application.AppStateListener;
 import org.sonar.process.ProcessId;
 
-public class AppStateClusterImpl implements AppState {
-  static final String OPERATIONAL_PROCESSES = "operational_processes";
-  private static final String LEADER = "leader";
+import static org.sonar.application.config.SonarQubeVersionHelper.getSonarqubeVersion;
+
+public class AppStateClusterImpl implements AppState, AutoCloseable {
```
Comment (SonarTech) [inline server/sonar-process-monitor/src/main/java/org/sonar/application/cluster/AppStateClusterImpl.java]: MINOR "AutoCloseable" is already extended by "AppState"; there is no need to implement it here. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1939)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java (hunk 2 above)
Comment (ehartmann) [inline server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java]: Not applicable for this use case.

File: server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+package org.sonar.application.config;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.charset.Charset;
+
+public class SonarQubeVersionHelper {
+  private static final String SONARQUBE_VERSION_PATH = "/sonarqube-version.txt";
```
Comment (SonarTech) [inline server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java]: MINOR Refactor your code to get this URI from a customizable parameter. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1075)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: server/sonar-process-monitor/pom.xml [hunk 6]
```diff
@@ -60,4 +69,14 @@
       <scope>test</scope>
     </dependency>
   </dependencies>
+
+  <build>
+    <resources>
+      <resource>
+        <!-- Used to resolve variables in files sq-version.txt and sonar-api-version.txt -->
```
Comment (simonbrandhof) [inline server/sonar-process-monitor/pom.xml]: Incorrect filenames.

File: server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  public static String getSonarqubeVersion() {
+    if (sonarqubeVersion == null) {
+      loadVersions();
+    }
+    return sonarqubeVersion;
+  }
+
+  private static synchronized void loadVersions() {
+    try {
+      try (BufferedReader in = new BufferedReader(new InputStreamReader(SonarQubeVersionHelper.class.getResourceAsStream(SONARQUBE_VERSION_PATH), Charset.forName("UTF-8")))) {
```
Comment (simonbrandhof) [inline server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java]: `Charset.forName("UTF-8")` could be replaced by `StandardCharsets.UTF_8`

File: server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        sonarqubeVersion = in.readLine();
+      }
+    } catch (IOException e) {
+      throw new IllegalStateException(
+        String.format("Cannot load %s from classpath", SONARQUBE_VERSION_PATH),
+        e
```
Comment (simonbrandhof) [inline server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java]: FYI we are used to put these parameters on the same line. String.format can be "static imported":
``` 
throw new IllegalStateException(format("Cannot load %s from classpath", SONARQUBE_VERSION_PATH), e);
```

File: server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java [hunk 9]
```diff
@@ -40,14 +44,19 @@
...
+import static org.sonar.application.config.SonarQubeVersionHelper.getSonarqubeVersion;
 
 public class AppStateClusterImplTest {
 
   @Rule
   public ExpectedException expectedException = ExpectedException.none();
 
   @Rule
-  public TestRule safeGuard = new DisableOnDebug(Timeout.seconds(10));
+  public TestRule safeGuard = new DisableOnDebug(Timeout.seconds(20));
```
Comment (simonbrandhof) [inline server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java]: why this change? Can the test be longer than 10 seconds?

File: server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java [hunk 10]
```diff
@@ -116,6 +125,56 @@ public void simulate_network_cluster() throws InterruptedException {
     }
   }
 
+  @Test
+  public void appstateclusterimpl_must_set_sonarqube_version() {
+    TestAppSettings settings = new TestAppSettings();
+    settings.set(ProcessProperties.CLUSTER_ENABLED, "true");
+    settings.set(ProcessProperties.CLUSTER_NAME, "sonarqube");
```
Comment (simonbrandhof) [inline server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java]: The 3 above lines should be replaced by `newClusterSettings()`

File: server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java (hunk 9 above)
Comment (ehartmann) [inline server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java]: Yes unfortunately, this test is launching 2 cluster and its duration is ~10s. 
I was thinking to move it under ITs, do you agree ?

File: server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+
+import static java.lang.String.format;
+
+public class SonarQubeVersionHelper {
+  private static final String SONARQUBE_VERSION_PATH = "/sonarqube-version.txt";
```
Comment (SonarTech) [inline server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java]: MINOR Refactor your code to get this URI from a customizable parameter. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1075)

File: server/sonar-process-monitor/src/main/java/org/sonar/application/cluster/AppStateClusterImpl.java [hunk 12]
```diff
@@ -99,12 +105,18 @@ public AppStateClusterImpl(AppSettings appSettings) {
       // Use slf4j for logging
       .setProperty("hazelcast.logging.type", "slf4j");
 
+    hzConfig.getMemberAttributeConfig()
```
Comment (simonbrandhof) [inline server/sonar-process-monitor/src/main/java/org/sonar/application/cluster/AppStateClusterImpl.java]: as discussed this attribute is not used and should be dropped.

File: server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java (hunk 9 above)
Comment (simonbrandhof) [inline server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java]: The "m" is missing :-) Duration should be ~10ms !
Let's try the following approach:
* extract AppState#registerSQVersion(String). Throws ISE if version different than the first registered version.
* App calls appState.registerSQVersion(loadSQVersion())
* AppStateClusterImplTest does not need to start a Hazelcast cluster. It calls twice registerSQVersion().

File: server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java [hunk 13]
```diff
@@ -116,6 +121,41 @@ public void simulate_network_cluster() throws InterruptedException {
     }
   }
 
+  @Test
+  public void appstateclusterimpl_must_set_sonarqube_version() {
```
Comment (simonbrandhof) [inline server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java]: name should be something like `registerSonarQubeVersion_publishes_version_on_first_call()`

File: server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java [hunk 14]
```diff
@@ -116,6 +121,41 @@ public void simulate_network_cluster() throws InterruptedException {
...
+    TestAppSettings settings = newClusterSettings();
+
+    try (AppStateClusterImpl appStateCluster = new AppStateClusterImpl(settings)) {
+      appStateCluster.registerSonarQubeVersion(getSonarqubeVersion());
```
Comment (simonbrandhof) [inline server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java]: no need to use `getSonarqubeVersion()`. It should be replaced by a fake value like "1.0".

File: server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java [hunk 15]
```diff
@@ -116,6 +121,41 @@ public void simulate_network_cluster() throws InterruptedException {
...
+      HazelcastInstance hzInstance = createHazelcastClient(appStateCluster);
+      assertThat(hzInstance.getAtomicReference(SONARQUBE_VERSION).get())
+        .isNotNull()
+        .isInstanceOf(String.class)
+        .isEqualTo(getSonarqubeVersion());
+    }
+  }
+
+  @Test
+  public void incorrect_sonarqube_version_must_trigger_an_exception() throws IOException, InterruptedException, IllegalAccessException, NoSuchFieldException {
```
Comment (simonbrandhof) [inline server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java]: we usually name such test `registerSonarQubeVersion_throws_ISE_if_initial_version_is_different()`. Note that `throws Exception` is simpler.

File: server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java [hunk 16]
```diff
@@ -116,6 +121,41 @@ public void simulate_network_cluster() throws InterruptedException {
...
+    settings.set(ProcessProperties.CLUSTER_ENABLED, "true");
+    settings.set(ProcessProperties.CLUSTER_NAME, "sonarqube");
+
+
+    try (AppStateClusterImpl appStateCluster = new AppStateClusterImpl(settings)) {
+      // Register first version
+      appStateCluster.registerSonarQubeVersion("1.0.0");
+
+      expectedException.expect(IllegalStateException.class);
+      expectedException.expectMessage(Matchers.matches("The local version .* is not the same as the cluster 1\\.0\\.0"));
```
Comment (simonbrandhof) [inline server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java]: Why not simply `expectedException.expectMessage("The local version 2.0.0 is not the same as the cluster 1.0.0");` ?

File: server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+package org.sonar.application.config;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+
+import static java.lang.String.format;
+
+public class SonarQubeVersionHelper {
```
Comment (simonbrandhof) [inline server/sonar-process-monitor/src/main/java/org/sonar/application/config/SonarQubeVersionHelper.java]: Missing SonarQubeVersionHelperTest !

File: server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java (hunk 15 above)
Comment (ehartmann) [inline server/sonar-process-monitor/src/test/java/org/sonar/application/cluster/AppStateClusterImplTest.java]: My bad, auto adding of exception. I'll fix this.
